extmod/modopenamp.c

```diff
@@ -140,6 +144,12 @@ static int endpoint_recv_callback(struct rpmsg_endpoint *ept, void *data, size_t
     endpoint_obj_t *self = metal_container_of(ept, endpoint_obj_t, ep);
     if (self->callback != mp_const_none) {
         mp_call_function_2(self->callback, mp_obj_new_int(src), mp_obj_new_bytearray_by_ref(len, data));
+    } else {
+        for (size_t i=0; i<len; i++) {
+            if (ringbuf_put(&self->ringbuf, ((uint8_t *) data)[i]) == -1) {
```

Comment:
The problem with a ring buffer used in this way is that the boundaries of the message are lost.

The behaviour of the endpoint should match UDP sockets, and a recv/read should return a single packet/message.  Eg if 2 messages are pending on the endpoint then the code needs to do 2 separate reads to get them both.

Reply:
Yeah I thought about it, but what if you call `recv` with a size that's less than the size of the next pending message ?

Reply:
Then the message is truncated to fit the requested number of bytes, and the rest of that message is discarded (that's how UDP recv/recvfrom work).

Reply:
Okay, updated. I've also found a way to avoid double copying, see comment message.